### DISCUSSION TITLE
Add GoogleCalendarEvent class with "attendees" attribute

### DISCUSSION THREAD

**Discussion by sfdye** [home-assistant/architecture#1099] [Entity Models]
Similar to https://github.com/home-assistant/architecture/discussions/884

Draft PR: https://github.com/home-assistant/core/pull/119457

## Use case

Taking a step back, my use case is that I want to get an notification for any incoming calendar events on a specific calendar, which 
- I have accepted
- Has zoom link(s)
- Not "all day" event
- Within my working hours

This is the automation I created, which almost works except that it will give false positives for non-accepted events.

![CleanShot 2024-06-13 at 07 05 33](https://github.com/home-assistant/architecture/assets/1016390/0ed39c68-3128-4ad0-8110-c5feab902c23)

I'd be happy to drop this proposal and PR if the goal can be achieved this without adding the `attendees` attribute.

Similar FR is raised here:
https://community.home-assistant.io/t/google-calendar-invitation-acceptance-status-as-attribute/429269

## Google Calendar API

The `attendees` attribute is exposed by GCal API as well as by [gcal_sync](https://github.com/allenporter/gcal_sync/blob/2456f6a69f56d0cc895885f5885a4e61789399b4/gcal_sync/model.py#L563).

![CleanShot 2024-06-13 at 07 11 32](https://github.com/home-assistant/architecture/assets/1016390/2d2b1257-b82a-478c-9fb6-5decbb3234a1)

The current method I am thinking to "filter" by accepted events is to see whether the current user is in the `attendees` list with `responseStatus` as `accepted` (or maybe even `tentative`)

![CleanShot 2024-06-13 at 07 16 20](https://github.com/home-assistant/architecture/assets/1016390/b592c63e-d389-40a8-86dc-587d895766cb)

I am open to explore alternatives without any code change 😄

**Comment by sfdye** [comment]
> (2) the integrations that we plan to make support this. just google calendar probably isn't enough

@allenporter sorry can I elaborate? First time contributing to HA, so bear with me. Do you mean other google related integrations or?

**Reply by allenporter** [reply]
Attendees is part of rfc5545 so adding it can make sense. Given that the proposal is to add this to the calendar event API, we'd want to look at all the integrations that could benefit from this. The main potential candidate I can think of would be CalDAV but there may be others (e.g. todoist). Not sure this makes sense for local calendar. Would we show anything in the UI for this?

**Reply by sfdye** [reply]
Interesting, it looks like CalDav supports `attendees` as well ([rfc6638](https://datatracker.ietf.org/doc/rfc6638/))

**Reply by sfdye** [reply]
> Not sure this makes sense for local calendar

Does not make a lot of sense for local calendar, but we can argue sole attendee == creator

**Reply by sfdye** [reply]
> but there may be others (e.g. todoist)

Todoist (I used for a few years before) definitely needs this, as it supports sharing

**Reply by sfdye** [reply]
> Would we show anything in the UI for this?

To be honest, I don't think this is really needed for automation purposes. I tend to just hide it, if possible.

**Reply by allenporter** [reply]
How would the attendees and statuses be exposed as attributes? I think we typically don't have attributes that are lists of key/value pairs like attendees essentially is. A raw dict of email addresses and statuses doesn't feel quite right to expose at this level.

Given the constrained use case, it implies that we probably don't need to expose status for all participants, just the owners as you mention. (If we can avoid trying to expose identities that seems helpful. We could explore mapping attendees to home assistant persons, though that seems a bit much to take on and is not part of the use case provided.)

So i see two options:
(A) We expose "participant status" for the owner. This could allow either exposing in automations or exposing in the UI to filter for visibility.  Integration decides who owner is. 
(B) We make integration specific options to filter out  declined events. 

I think to your original question: Does it ever make sense to surface declined events as the state for the binary sensor? I think the answer is probably "no" which means (B) and the integration should filter these out with no option to show them. This would be a breaking change. Users would lose ability to have automations on declined events.  This would mean we can't tell the difference between accepted vs no participant status.

**Reply by sfdye** [reply]
I prefer (B) because
1. less effort to implement 
2. less blast radius 
3. backward compatible (default: include declined events with option to exclude)
wdyt?

**Reply by allenporter** [reply]
Perhaps we may want to "rip the bandaind" and make a breaking change to skip declined events with an option to include in the configuration options. WDYT?

**Reply by sfdye** [reply]
LGTM. FWIW, I think excluding declined events should be the correct default behavior. This breaking change should be minor. I can't think of any legit use cases to automate upon declined events.

**Reply by sfdye** [reply]
I will be travelling soon so may only get back to this next week only. In the meantime, anyone feel free to "rip the bandaid" 😄

**Comment by sfdye** [comment]
On this note, actually I don't understand why by default all events (even declined) events are included.

**Reply by sfdye** [reply]
I get it now, it requires the least amount of effort to not filter out anything 😄

**Comment by Quentame** [comment]
On architecture discussions, you need to suggest modifications available for most integrations of a domain or a platform.

You should search for all integrations that can benefit from your suggestion, and not be specific to only one.

Actually, yours should be applicable to almost all integrations with a calendar platform.

**Comment by Wetzel402** [comment]
Bringing my `attendee` discussion over here to unify the discussion under one thread...

## Proposal 

**HA Core:**
- Add `attendees` to `calendar` component in a unintrusive way. 
- Add `attendees` to `local_calendar` component

**HA Frontend:**
- Map `person` to checkbox in create/edit event dialog
- Improve event details dialog and add `attendees`

## Screenshots
![image](https://github.com/user-attachments/assets/51d9ee76-73e4-4ebd-993b-ec20587479eb)
![image](https://github.com/user-attachments/assets/e2cf580f-33cb-432f-becf-f1f8de45f133)

## Use case
- Current draft PRs provide a local replacement for cloud based family calendars (e.g. [cozi](https://www.cozi.com), [family wall](https://www.familywall.com/en/index.html), etc.)
- Future PRs could harness the changes in `calendar` to add `attendees` for other calendar components (e.g. google calendar, CalDAV, etc)

## Architecture 
- Changes in `calendar` are minimal leaving the implementation of `attendees` up to the individual integration
- Changes in `local_calendar` and `frontend` tie `attendees` to the `person` within HA by name
  - Each `attendee` receives a `common_name` from the `person` `name`
  - `ical` requires a `uri` but HA currently does not have an email associated with a `person`.  This is currently left as an empty string. 

## Future
- Implement import of ics for `local_calendar`
- Ability to tie `person` to `todo` by name

## Other considerations 
- Add email to `person`
- Add `notify[]` array to `person` similar to `device_tracker`


Related to...
https://github.com/home-assistant/architecture/discussions/1099
https://github.com/home-assistant/architecture/discussions/884

Draft PRs...
https://github.com/home-assistant/core/pull/122561
https://github.com/home-assistant/frontend/pull/21471

**Comment by Wetzel402** [comment]
Thoughts posed by @allenporter in the other thread...

> People
> You're going to need to further expand on the use cases for people quite a bit. This is going to get complex (more than is specified here) so its going to need significant motivation to be accepted.
> 
> Once we have more use cases settled, attendees is still underspecified here. Some thoughts are:
> 
> 
> - How will these attendees integrate in the rest of home assistant? For example: used in automations, or how does it integrate with other ways things are associated with people like people + locations

My PRs tie `person` in by name only currently so I see your point.  Based on that an automation would be difficult to build.  Do we expand on `person` first by adding email and/or notify[]?  Do we add `person[]` to `event` and if so how does that get handled?
> - You seem to be proposing an attendee object with a single field, but i can't tell, so we'll want to discuss what that object is. [Attendees](https://www.rfc-editor.org/rfc/rfc5545#section-3.8.4.1) are a very rich "object" with lots of fields so what's in and whats out now and going forward

In my PR the changes in `calendar` are minimal with the changes in `local_calendar` handling all the details. 
https://github.com/home-assistant/core/pull/122561/files#diff-a1473c45f850b58510c6b9371eb1126416908d5a3228fdf6abe677fd15dc6cb2
> - More on the semantics of what these fields mean in terms of home assistant -- i'm not understanding if the proposal is exact name matching, or entities, or what.

Currently it's matching to a name in the `frontend`.  Without adding unique emails to `person` I don't see another option currently. 
> - More conviction about how multiple integrations can integrate with this in an understandable way (e.g. how are non-person attendees handled). We shouldn't punt on having a credible plan on how multiple integrations will support this.

Are you referring to beyond calendar integrations?  Currently the obvious two would be calendar and todo.  

As for non-person attendees, that isn't possible under my current PR for `local_calendar` but I understand the concern and use case.  I will say that other family calendars I've used do not provide such a provision either.  You add your family to the service and they are either attending or not.  There isn't an option for non family attendees.  With that said, do we handle it that way for `local_calendar` where only a `person` can attend?  If not, how do we handle the ability for non-`person` to attend?
 Google Calendar currently can't create/edit an event from HA so for that integration attendees could be parsed without a match to `person` which would allow for non-`person` `attendees` or we attempt to match to `person` but still add any non-`person` to `attendees`.  

This topic is definitely much more complex than I originally anticipated.

**Reply by Wetzel402** [reply]
To expand on `local_calendar` making use of `person` as `attendees`...
- Allows for the replacement of cloud family calendars with a fully local option
- automation...
   - `notify` `person` of `event` due or `event` creation
   - fire automation based on event and make use of `attendees`
   - create/edit/delete event based on `person`
- some example use cases...
   - if `local_calendar` contains an `event` and `person` is sleeping we could turn on/increase brightness of lights, `notify` the `person`, open blinds/curtains, etc
   - if `local_calendar` contains an 'event' and `person` is not there we could `notify` other `person` (e.g. if my child is missing their appointment notify me)
   - if `event` `location` is "zoom" and the `attendee` is me prepare my home office for the zoom meeting by turning on the lights, waking my PC, etc
   - if `person` enters `event` location update/delete/create event, trigger further automation notifying other `attendees`, etc
   - could be used by HA for reminders (e.g. a reminders calendar where events would trigger a notification for the `attendees`)  There are many work arounds for this problem but no official reminder support that I'm aware of.   This could further be harnessed by HA through `assist`. 

Some of these also rely on `location` which is not yet fully implemented.  See https://github.com/home-assistant/architecture/discussions/1119

**Reply by allenporter** [reply]
This didn't really answer the original questionas, and not looking to get into PR details here.  The proposal is vague about what the attendees object actually is, leaving it up to the integration local calendar to decide, but we're looking for something that works with other calendar integrations.  This needs to have a well form proposed spec that has a credible path to work with multiple integrations and clear semantics for how it could be used for automations, etc.  I don't know how to represent a person and that needs to be answered.

**Reply by Wetzel402** [reply]
I follow.  An individual attendee is documented by your `CalAddress`  type and could be implemented with things like Google Calendar and CalDAV but the complication comes in when trying to tie that attendee to a person.  Do we tie it to a person or do we not?  If we do, maybe we need to have an arch discussion around adding an email address for `person` and using that to tie person and attendee.

As far as automations, I foresee exposing the attendees of an event so that the attendee can be useful. 

https://allenporter.github.io/ical/ical/types.html#CalAddress
